src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
     }
 }
 
+// If succeeded, U_SUCCESS(*pErr) should be true.
+// return -1 if couldn't borrow search handle from the SortHandle cache, otherwise, it return the slot number of the cache.
+static const int32_t GetSearchIteratorFromSortHandleUsingCollator(
+                        SortHandle* pSortHandle,
+                        const UCollator* pColl,
+                        const UChar* lpTarget,
+                        int32_t cwTargetLength,
+                        const UChar* lpSource,
+                        int32_t cwSourceLength,
+                        int32_t options,
+                        UErrorCode* pErr,
+                        UStringSearch** pSearchIterator)
+{
+    options &= CompareOptionsMask;
+    *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+
+    if (*pSearchIterator == NULL)
+    {
+        *pSearchIterator = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, pErr);
+        if (!U_SUCCESS(*pErr))
+        {
+            assert(FALSE && "Couldn't open the search iterator.");
+            return -1;
+        }
+
+        UStringSearch* pNull = NULL;
+        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], *pSearchIterator, pNull))
```

Comment:
Did you consider instead just having each thread store a single cached handle in a [ThreadStatic]?

Reply:
Honestly no but I can see some issues doing that: 

- This will require a dictionary lookup with every call.
- If there are many created threads means more native allocation will occur and this may be a concern especially we don't free them.
- minor: Threadstatic still have some extra cost and we'll be required to pass extra parameter to the pinvoke calls. 

If you think it is worth to try, I can try prototype it but I am not sure the concerns I have with it is mitigated.